tft.py

```diff
@@ -0,0 +1,189 @@
+import json
+import glob
+import mmap
+import copy
+import re
+import os
+from json import JSONEncoder
+from cdragontoolbox.binfile import BinFile, BinHashBase, BinHashValue
+
+
+class NaiveJsonEncoder(JSONEncoder):
+    def default(self, other):  # pylint: disable=method-hidden
+        if isinstance(other, BinHashBase):
+            if other.s is None:
+                return other.hex()
+            return other.s
+        return other.__dict__
+
+
+class TFTTransformer:
+    def parse(self, input):
+        map22_file = os.path.join(input, "data", "maps", "shipping", "map22", "map22.bin")
+        character_folder = os.path.join(input, "data", "characters")
+
+        map22 = BinFile(map22_file)
+        items = self.parse_items(map22)
+        traits = self.parse_traits(map22)
+        champs = self.parse_champs(map22, traits, character_folder)
+
+        # clean up data for JSON export
+        [x.pop("internal") for x in traits]
+        [x.pop("internal") for x in items]
+
+        self.data = {"items": items, "traits": traits, "champs": champs}
+
+    def export(self, output, font_files=[], output_template=False):
+        os.makedirs(output, exist_ok=True)
+        template = json.dumps(self.data, cls=NaiveJsonEncoder, indent=4, sort_keys=True)
+
+        if output_template:
+            with open(os.path.join(output, "template.json"), "w", encoding="utf-8") as f:
+                f.write(template)
+
+        for language in font_files:
+            use = template
+            code = os.path.basename(language)[11:-4]
+
+            with open(language, "r", encoding="utf-8") as f:
+                for line in f:
+                    if line.startswith("tr"):
+                        key, val = line.split("=", 1)
+                        key = key[4:-2]
+                        val = val[2:-2]
+                        use = use.replace(key, val)
+
+            with open(os.path.join(output, code + ".json"), "w", encoding="utf-8") as f:
+                f.write(use)
+
+    def parse_items(self, map22):
+        item_collection = [x for x in map22.entries if x.type == "TftItemData"]
+        items = []
+
+        for item in item_collection:
+            name = item.getv("mName")
+            if "Template" in name or name == "TFT_Item_Null":
+                continue
+
+            effects = []
+            for effect in item.getv("effectAmounts") if item.get("effectAmounts") else []:
+                effects.append({"name": effect.getv("name"), "value": effect.getv("value")})
+
+            items.append(
+                {
+                    "internal": {"hex": item.path.hex(), "name": item.getv("mName")},
+                    "id": item.getv("mId"),
+                    "name": item.getrv("c3143d66"),
+                    "desc": item.getrv("765f18da"),
+                    "icon": item.getv("mIconPath"),
+                    "from": [x.hex() for x in item.getrv("8b83ba8a")] if item.getr("8b83ba8a") else [],
```

Comment:
Here you can rely on `get()` to return an empty list as default and avoid the ternary.
And you don't need to convert to a string, just use the integer hash.
```py
[x.h for x in item.getv(0x8b83ba8a, [])]
```